apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,336 @@
+        projectTokenManager.revokeVesting(_buyer, _purchaseId);
+
+        // (Token manager) ~~~> project tokens ~~~> (burn 💥)
+        projectTokenManager.burn(address(projectTokenManager), tokensSold);
+
+        emit TokensRefunded(_buyer, daiToRefund, tokensSold, _purchaseId);
+    }
+
+    /**
+    * @notice Closes a sale that has reached the funding goal, sending raised funds to the fundraising pool and the beneficiary address, and initializes the Fundraising app by adding the raised funds as collateral tokens.
+    */
+    function close() public {
+        require(currentSaleState() == SaleState.GoalReached, ERROR_INVALID_STATE);
+
+        // (presale) ~~~> dai ~~~> (beneficiary)
+        uint256 daiForBeneficiary = totalDaiRaised.mul(percentFundingForBeneficiary).div(PPM);
+        require(daiToken.transfer(beneficiaryAddress, daiForBeneficiary), ERROR_DAI_TRANSFER_REVERTED);
+
+        // (presale) ~~~> dai ~~~> (pool)
+        uint256 daiForPool = daiToken.balanceOf(address(this));
+        require(daiToken.transfer(fundraisingPool, daiForPool), ERROR_DAI_TRANSFER_REVERTED);
+
+        saleClosed = true;
+
+        emit SaleClosed();
+    }
+
+    /*
+     * Getters
+     */
+
+    /**
+    * @notice Calculates the number of project tokens that would be obtained for `_daiAmount` dai tokens.
+    * @param _daiAmount uint256 The amount of dai tokens to be converted into project tokens.
+    */
+    function daiToProjectTokens(uint256 _daiAmount) public view returns (uint256) {
+        return _daiAmount.mul(daiToProjectTokenExchangeRate);
+    }
+
+    /**
+    * @notice Calculates the current state of the sale.
+    */
+    function currentSaleState() public view returns (SaleState) {
+        if (startDate == 0) {
+            return SaleState.Pending;
+        } else if (totalDaiRaised >= daiFundingGoal) {
+            if (saleClosed) {
+                return SaleState.Closed;
+            } else {
+                return SaleState.GoalReached;
+            }
+        } else if (_timeSinceFundingStarted() < fundingPeriod) {
+            return SaleState.Funding;
+        } else {
+            return SaleState.Refunding;
+        }
+    }
+
+    /*
+     * Internal
+     */
+
+    function _timeSinceFundingStarted() private returns (uint64) {
+        if (startDate == 0) {
+            return 0;
+        } else {
+            return getTimestamp64().sub(startDate);
+        }
+    }
+
+    function _calculateExchangeRate() private {
+        uint256 exchangeRate = daiFundingGoal.mul(PPM).div(CONNECTOR_WEIGHT_PPM).mul(percentSupplyOffered).div(PPM);
+        daiToProjectTokenExchangeRate = exchangeRate;
+    }
+
+    function _setProjectToken(MiniMeToken _projectToken, TokenManager _projectTokenManager) private {
+        require(isContract(_projectTokenManager), ERROR_INVALID_TOKEN_CONTROLLER);
+        require(_projectToken.controller() != address(projectTokenManager), ERROR_INVALID_TOKEN_CONTROLLER);
```

Comment:
A better check would be `projectTokenManager.token() == address(_projectToken)`.

The token may have had its controller assigned to the Token Manager, but it's not necessarily the case that the Token Manager has been initialized with that token.